specklesystems/speckle-sharp
# feat(dui2/rhino/revit): better reporting and previews

## Description

This is a major PR that includes changes in:

**Core** 
- new `ApplicationObject`: this is scoped in [notion](https://www.notion.so/speckle/ApplicationObject-6cfbb143a22844378f3fbd4047bc8f67) and replaces the old `ApplicationPlaceholderObject` class. **Breaking change** Old files with saved streams will not deserialize correctly and would have to be re-saved

**DUI2**
- moved `Report` to a tab in `StreamEditView`: this will display old report logging if no `ApplicationObjects` are saved in `ProgressViewModel` or display a list of `ApplicationObjects` if they are present
- new connector bindings: `PreviewSend`, `PreviewReceive`, `ResetDocument`, and an updated `SelectClientObjects`
- [x] Todo: add copy report button that summarizes all `ApplicationObjects`
- [x] Todo: preview button linked to send/receive button enabled state;
- [ ] Todo: add operation errors to top of report
- [ ] Todo: add progress bar on preview receive

**Rhino**
- Refactored Receive method quite heavily, including the `FlattenCommitObject` method to work with the new `ApplicationObject` for reporting
- Implemented `PreviewSend` and `PreviewReceive`
Known issues: 
- [x] previewing blocks and views, which are created during conversion. Should be resolved with new block helper method and skipping views.

**Revit**
- **IMPORTANT** Refactored conversions to only return a single `ApplicationObject` instead of a List - the old nested elements should be stored in the `converted` and `createdIds` prop of the parent `ApplicationObject`
- Updated report for receive
Known issues: 
- [x] update report for send as well
- [x] check structural conversions as some of these didn't quite make sense with how they were handling the old `ApplicationPlaceholderObject`

**ETABS/CSI/Tekla Structures**
Known Issues:
- [x] report not scrollable
- [x] preview recieve crashes plugin 

Other sharp connectors
- added new connector bindings

> Next steps include potentially refactoring the `ReportObjects` property on `ProgressReport` (which is used to pass `ApplicationObjects` between the connector and converter) and the `ContextObjects` connector property, as there seems to be some overlap of functionality.
Known bugs:
clicking preview receive on and off quickly losing sync state
switching preview receive on and off wipes report (this should be preserved)

- Fixes #1447
- Fixes #1448
- Fixes #1449


## Type of change

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How has this been tested?

- Manual Tests: Rhino sends and receives
- Requires more thorough testing to check for build errors and breaking behavior in:
- [x] Revit @teocomi 
- [x] Structural connectors @Reynold-Chan 
- [x] Bentley

## Docs

- Will be updated ASAP
